pennylane/math/ode.py

```diff
@@ -0,0 +1,246 @@
+        y0 (ndarray): initial value ``y(t0) = y0``
+        ts (ndarray): finite time steps for the ODE solver to take.
+
+    .. note::
+
+        Unlike many standard implementations, this ``odeint`` solver does not adaptively choose the step
+        sizes, but rather a fix list of times for evaluation have to be provided in ``ts``. For an adaptive step
+        size solver, we refer to `jax.experimental.ode.odeint <https://github.com/google/jax/blob/main/jax/experimental/ode.py>`_.
+
+    .. warning::
+
+        This function is written for ``jax`` only and will not work with other machine learning frameworks typically encountered in PennyLane.
+
+    **Example**
+
+    Let us look at the very simple initival value problem $dy/dt = y,$ $y(0)=3$.
+    The analytic solution is $y(t) = 3e^t$. Say we are interested in the solution at
+    time $t=2$, i.e. $y(2)=3e^2$. We can compute this numerically using ``odeint``
+
+    .. code:: python3
+
+        t = jnp.linspace(0, 2, 20)
+
+        def fun(y, t):
+            return y
+
+        y0 = jnp.array((3.))
+        res = qml.math.odeint(fun, y0, t)
+
+    >>> print(qml.math.isclose(res, 3*jnp.exp(2)))
+    True
+
+    As long as ``fun(y, t)`` outputs arrays of the same shape as ``y``, we can use arbitrary input
+    shapes for ``y0``.
+
+    >>> y0 = jnp.ones((5), dtype=float)
+    >>> y1 = qml.math.odeint(fun, y0, t)
+    >>> print(y0.shape, y1.shape)
+    (5,) (5,)
+
+    We can solve the time-dependent Schrodinger equation for the unitary evolution operator
+
+    .. math:: \frac{d}{dt}U = -i H(t) U
+
+    for a time-dependent Hamiltonian :math:`H(t) = X_0 X_1 + v \sin(t) Z_0 Y_1`
+    in the time window ``(t0, t1) = (0, 4)`` using ``odeint`` in the following way:
+
+    .. code:: python3
+
+        XX = qml.matrix(qml.PauliX(0) @ qml.PauliX(1))
+        ZY = qml.matrix(qml.PauliZ(0) @ qml.PauliY(1))
+
+        ts = jnp.linspace(0., 2., 50)
+        y0 = jnp.eye(2**2, dtype=complex)
+        params = jnp.ones(1, dtype=complex)
+
+        def func(y, t, params):
+            H = jnp.array([XX + params[0] * jnp.sin(t) * ZY])
+            return -1j * jnp.sum(H, axis=0) @ y
+
+        U = qml.math.odeint(func, y0, ts, params)
+
+    Formally, this solution can be written as the time-ordered exponential
+    :math:`U(t_0, t_1) = \text{Texp}\left[-i \int_{t_0}^{t_1}d\tau H(\tau)\right]`
+    (see `Dyson Series <https://en.wikipedia.org/wiki/Dyson_series>`_).
+
+    Note that this computation can be backward differentiated with respect to the parameters ``params`` via
+
+    >>> jac = jax.jacobian(qml.math.odeint, argnums=3, holomorphic=True)(func, y0, ts, params)
+    """
+    y0, unravel = ravel_pytree(y0)
+    func = ravel_first_arg(func, unravel)
+    out = _odeint(func, y0, ts, atol, rtol, *args)
+    return unravel(out)
+
+
+def _abs2(x):
+    """Computing the squared absolute value of x"""
+    if jnp.iscomplexobj(x):
+        return x.real**2 + x.imag**2
+    return x**2
```

Comment:
Just a suggestion, but could we maybe use `abs(x) ** 2` instead of this `if` statement?

Filippo said that `if` statements are very slow in jax

Reply:
I am happy with `jnp.abs(x)**2` though take into account that it does an unnecessary sqrt and then squaring!